plutus-contract/test/Spec/TxConstraints/MustSpendScriptOutput.hs

```diff
@@ -43,284 +51,365 @@ tests =
-            awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx4
-
-    in checkPredicateOptions
-        defaultCheckOptions
-        "Phase-2 validation failure when onchain mustSpendScriptOutput constraint expects a different redeemer"
-        (assertFailedTransaction (\_ err _ -> case err of {Ledger.ScriptFailure (EvaluationError ("L8":_) _) -> True; _ -> False }))
+        (assertFailedTransaction (\_ err _ -> case err of {L.ScriptFailure (EvaluationError ("L8":_) _) -> True; _ -> False }))
         (void $ trace contract)
 
--- | Uses onchain and offchain constraint mustSpendScriptOutputWithMatchingDatumAndValue to spend a UTxO locked by the script with matching datum and value
+-- | Phase-2 validation failure only when V2 script using onchain mustSpendScriptOutput constraint
+-- expects a different redeeemer with V2+ script
+phase2ErrorOnlyWhenMustSpendScriptOutputUsesWrongRedeemerWithV2Script :: TestTree
+phase2ErrorOnlyWhenMustSpendScriptOutputUsesWrongRedeemerWithV2Script =
+    testGroup "phase2 error only when mustSpendScriptOutput uses wrong redeemer with V2+ script"
+
+    [ checkPredicateOptions
+      defaultCheckOptions
+      "No phase-2 validation failure when V1 script using onchain mustSpendScriptOutput constraint expects a different redeemer"
+      (valueAtAddress someAddress (== utxoValue)
+      .&&. assertValidatedTransactionCount 2)
+      (void $ trace $ mustSpendScriptOutputsContract' PlutusV1 7 6 False)
+
+    , checkPredicateOptions
+      defaultCheckOptions
+      "Phase-2 validation failure when V2 script using onchain mustSpendScriptOutput constraint expects a different redeemer"
+      (assertFailedTransaction (\_ err _ -> case err of {L.ScriptFailure (EvaluationError ("L8":_) _) -> True; _ -> False }))
+      (void $ trace $ mustSpendScriptOutputsContract' PlutusV2 5 5 False)
+    ]
+
+-- | Uses onchain and offchain constraint mustSpendScriptOutputWithMatchingDatumAndValue to spend a
+-- UTxO locked by the script with matching datum and value
 validUseOfMustSpendScriptOutputWithMatchingDatumAndValue :: TestTree
 validUseOfMustSpendScriptOutputWithMatchingDatumAndValue =
     let nScriptOutputs  = 5
         scriptOutputIdx = nScriptOutputs - 1
     in checkPredicateOptions
         defaultCheckOptions
         "Successful use of mustSpendScriptOutputWithMatchingDatumAndValue to spend a UTxO locked by the script with matching datum and value"
-        (valueAtAddress mustSpendScriptOutputWithMatchingDatumAndValueScrAddress (== Value.scale 4 utxoValue)
+        (valueAtAddress someAddress (== V.scale 4 utxoValue)
         .&&. assertValidatedTransactionCount 2)
-        (void $ trace $ mustSpendScriptOutputWithMatchingDatumAndValueContract nScriptOutputs (scriptOutputIdx, scriptOutputIdx) (utxoValue, utxoValue))
+        (void $ trace $ mustSpendScriptOutputWithMatchingDatumAndValueContract PlutusV1 nScriptOutputs (scriptOutputIdx, utxoValue) (scriptOutputIdx, utxoValue))
```

Comment:
Do we want to duplicate the test cases for PlutusV2? Basically copy-paste the `TestTree` function and change `PlutusV1` to `PlutusV2`.

Reply:
Yeah, that is the case now but @berewt is currently working on addressing this by combining use of his `TestContext` with this `Language` to run tests like this for each supported plutus version without duplication of `TestTree`. Should be incoming soon.

Reply:
Yes, it was part of my preparatory work for PLT-659. We can either wait for its completion or I can add it to this PR, as you prefer (the first solution would save me some time though).

Reply:
Cool :)